SMA speedwire support for inverters

bboehmke opened the pull request:
This change adds support to retrieve values from SMA inverters via the speedwire protocol.

I used a library I wrote some time ago that implements the SMA speedwire protocol in go:
[gitlab.com/bboehmke/sunny](https://gitlab.com/bboehmke/sunny)
The library is far from perfect and still not complete but it provides all values that are used by evcc.


This configuration
```yaml
meters:
  - name: grid
    type: sma
    uri: 192.168.4.50
  - name: pv
    type: sma
    uri: 192.168.4.51
  - name: battery
    type: sma
    uri: 192.168.4.52
```
results in this values:
```
battery
-------
Power: -684W
SoC:   14%

grid
----
Power:          -6W
Current L1..L3: 3.01A 1.57A 1.44A

pv
--
Power:          1166W
Current L1..L3: 1.74A 1.73A 1.74A
```

There are still some open points to solve but before I invest to much time I would ask for a first feedback.

TODO:
- [x] select device by serial number
- [x] ~support custom power value~
- [x] support energy value
- [x] ~update test~ removed
- [x] cleanup old SMA implementation?
- [x] detect task
- [x] ~support devices on multiple network interfaces~ (not supported by old implementation)

bboehmke committed:
first draft of SMA speedwire support for inverters

andig commented:
/cc @DerAndereAndi

andig commented on meter/sma.go:
```diff
@@ -27,62 +29,67 @@ type SMA struct {
-		powerO:  sma.Obis(power),
-		energyO: sma.Obis(energy),
-		recv:    make(chan sma.Telegram),
+	if password == "" {
```
Set default when decoding

bboehmke replied:
done

andig commented on meter/sma.go:
```diff
@@ -27,62 +29,67 @@ type SMA struct {
-			return nil, err
-		}
-		sma.Instance = instance
+	err := sunny.SetMulticastInterface(iface)
```
Iface can be nil or empty

bboehmke replied:
nil ist nicht möglich für einen string in go

aber für den leeren string haben ich eine Behandlung eingebaut

andig commented on meter/sma.go:
```diff
@@ -91,66 +98,97 @@ func NewSMA(uri, serial, iface, power, energy string) (api.Meter, error) {
+			sm.values.soc = sm.convertValue(soc)
+			sm.mux.Update()
+		} else {
+			// only exist for battery inverters
```
Remove else

bboehmke replied:
removed

andig commented on meter/sma.go:
```diff
@@ -174,11 +212,35 @@ func (sm *SMA) CurrentPower() (float64, error) {
+	switch v := value.(type) {
+	case float64:
+		return v
+	case int32:
```
Combine cases

bboehmke replied:
Das geht leider nicht weil dann der type switch bzw die `v` variable nicht mehr den type hat der im case steht

andig replied:
ups, ok :O

andig commented:
Please remember to fix the detect task as well. Is there any advantage innusing your lib ober the current implementation?

andig commented:
@bboehmke mir ist generell nicht klar, was Du mit dem PR eigentlich erreichen willst?

bboehmke commented:
Die bisherige Implementierung für das SMA speedwire protocol unterstützt nur die Home Manager aber nicht die Wechselrichter.
In meinem Fall möchte ich einen Batterie und einen PV Wechselrichter per Speedwire nutzen.

Da ich bereits eine Lib hatte die dies kann hab ich die bestehende Implementierung einfach ersetzt. 
Es ist natürlich auch eine Option die bestehende Implementierung zu erweitern.

DerAndereAndi commented:
Ist Speedwire bei deinen Geräten "besser" als ModBus?

andig commented:
> Die bisherige Implementierung für das SMA speedwire protocol unterstützt nur die Home Manager aber nicht die Wechselrichter.

ist das denn so? Kann gut sein, dass ich da einen Denkfehler habe (hab kein SMA)- was ist bei den anderen Geräten anders?

bitte nicht falsch verstehen- wenn uns das voran bringt sehr gerne, ich möchte es nur verstehen.

bboehmke commented:
@DerAndereAndi Ich hatte mir ModBus vor einiger Zeit mal angeschaut allerdings stand in der Doku das man höchstens alle 10s 5 Werte abfragen darf (bei Speedwire ist jede Sekunde kein Problem). Das war für meinen damaligen Usecase ([Live Anzeige der Werte](https://gitlab.com/bboehmke/sunny/-/raw/master/monitor_app.png)) etwas ungeeignet weshalb ich mich mit ModBus nicht weiter beschäftigt habe und auch nicht weiß ob das beide Wechselrichter beherrschen.

@andig Speedwire selbst kann verschiedene Sub-Protokolle enthalten. Die Energy Meter verwenden hier ein ganz anders Protokoll als die Wechselrichter. Ich hab dafür sogar mal nen [wireshark disector](https://gitlab.com/bboehmke/sunny/-/blob/master/wireshark/spw.lua) geschrieben.

Weitere vorteile die ich an Speedwire sehe:
* Auto discovery
* Keine Konfiguration an Wechselrichter notwendig

Natürlich gibt es auch Nachteile:
* Kein offizieller Support von SMA
* Protokoll nicht öffentlich

premultiply commented:
Die SMA-Geräte sprechen intern per Speedwire (UDP Unicast/Multicast) miteinander. In Gegensatz zu Modbus ist das immer aktiv und m.W.n auch bidirektional.

bboehmke committed:
some cleanup & current value fixes

DerAndereAndi commented:
@bboehmke magst evtl. zu uns in Slack kommen? Da lassen sich Verständnisfragen etc. leichter klären. Link ist auf der Wiki Seite.

bboehmke committed:
configure SMA device via serial number

bboehmke committed:
reworked detect task

bboehmke committed:
updated SMA interface library

bboehmke committed:
updated handling of energy value

DerAndereAndi committed:
Add scale to SMA meter for usage as PV meter

The SMA meters are and can be used as grid meter and PV meter. EVCC expects PV power to be positiv, but the meter would return it negativ as it is in export direction.

This is solved by adding an optional Scale property that is only used for the Power value. Default value for Scale is `1`

DerAndereAndi commented:
Hier muss noch die Ergänzung aus https://github.com/andig/evcc/pull/1133 hinzugefügt werden

andig commented on meter/sma.go:
```diff
@@ -27,130 +31,181 @@ type SMA struct {
 // NewSMAFromConfig creates a SMA Meter from generic config
 func NewSMAFromConfig(other map[string]interface{}) (api.Meter, error) {
 	cc := struct {
-		URI, Serial, Interface, Power, Energy string
```
wie wäre es power/energy drin zu lassen, aber mit WARNING im Logfile zu ignorieren? Sonst ist das ein breaking change bei dem evcc nicht mehr startet :O

bboehmke replied:
breaking change ist natürlich nicht schön.
c5cc2266 fügt die parameter und ein paar Warnungen hinzu

bboehmke committed:
added warning for SMA energy and power setting

bboehmke committed:
removed sma unit test and energy meter implementation

andig commented on detect/tasks/sma.go:
```diff
@@ -34,8 +36,9 @@ func SMAHandlerFactory(conf map[string]interface{}) (TaskHandler, error) {
 type SMAHandler struct {
 	mux      sync.Mutex
-	listener *sma.Listener
+	handled  bool
```
noch schöner wäre wohl das mit sync.Once zu machen, aber nicht kriegsentscheidend

bboehmke replied:
ist aber auch nicht ganz perfekt da sync.once ja eigentlich dafür gedacht einen Funktionsaufruf zu kapseln. 
Hier ist das ja die Methode die nur einmal aufgerufen werden soll (da sieht das mit sync.once glaube ich auch nicht schön aus)

andig replied:
yep- dann müssten die Innereien in eine Methode. Ich hänge nicht dran ;)

bboehmke replied:
Ich glaube das lasse ich erstmal so.
Vielleicht macht es sinn den detector für so etwas generell zu befähigen (braucht man bestimmt nicht nur für SMA)

andig commented on detect/tasks/sma.go:
```diff
@@ -60,46 +63,27 @@ func (h *SMAHandler) httpAvailable(ip string) bool {
-	if h.listener, err = sma.New(log, ""); err != nil {
+	devices, err := sunny.DiscoverDevices(h.Password)
+	if err != nil {
 		log.ERROR.Println("shm:", err)
```
s/shm/sma?

bboehmke replied:
shm hieß das bisher kann ich aber auch in sma ändern

bboehmke replied:
ich hab das jetzt alles von shm auf sma geänder (auch in den structs)

andig commented on meter/sma.go:
```diff
@@ -19,138 +21,196 @@ type values struct {
-		instance, err := sma.New(log, iface)
+	if iface != "" {
+		err := sunny.SetMulticastInterface(iface)
 		if err != nil {
```
bitte als Einzeiler: 

    if err :=... ; err!=nil ...

bboehmke replied:
done

andig commented on meter/sma.go:
```diff
@@ -19,138 +21,196 @@ type values struct {
-			sm.values.power = power
+	vals, err := sm.device.GetValues()
+	if err != nil {
+		sm.log.WARN.Printf("failed to get values: %v", err)
```
Error statt Warn?

bboehmke replied:
changed

andig commented on meter/sma.go:
```diff
@@ -19,138 +21,196 @@ type values struct {
+			sm.values.power = sm.convertValue(powerP) - sm.convertValue(powerM)
 			sm.mux.Update()
+		} else {
+			sm.log.WARN.Println("missing value for power")
```
dito?

bboehmke replied:
changed

andig commented on meter/sma.go:
```diff
@@ -19,138 +21,196 @@ type values struct {
+			sm.values.currentL3 = sm.convertValue(currentL3)
+			sm.mux.Update()
+		} else {
+			sm.log.WARN.Println("missing value for currentL3")
```
dito

bboehmke replied:
changed

andig commented on meter/sma.go:
```diff
@@ -19,138 +21,196 @@ type values struct {
+			sm.values.power = sm.convertValue(power)
+			sm.mux.Update()
+		} else {
+			sm.log.DEBUG.Println("missing value for power")
```
wozu?

bboehmke replied:
changed to error

andig commented:
Was mir noch auffällt: bei evcc kann ich oft mit `--log trace` auch die "raw" Gerätekommunikation mitschreiben. Das wäre optional, könnte ich mir aber als Erweiterung von sunny vorstellen zu Diagnosezwecken. Ich hänge aber nicht dran.

bboehmke committed:
more cleanup for sma meter

bboehmke committed:
renamed detector from shm to sma

andig commented:
LGTM: @DerAndereAndi ?

DerAndereAndi commented:
@andig bin noch am Testen

bboehmke committed:
updated sunny and added trace logger

bboehmke committed:
added Diagnose to sma meter

bboehmke committed:
fixed decorator

bboehmke committed:
handle SMA inverter by night

DerAndereAndi reviewed (approved):
LGTM 👍

premultiply commented:
Darf ich noch vorschlagen das Protokoll überall von "sma" in "speedwire" (analog zu modbus) umzubenennen?
Oder wie andi schreiben würde: `s/sma/speedwire` ;)

andig commented:
Dann müssten wir alle configs anpassen- insofern wohl nciht :(

premultiply commented:
Naja, da das Feature ja noch brandneu ist dürfte die Zahl der Betroffenen sehr überschaubar sein.

andig commented:
sma hatten wir schon länger, nix neu

premultiply commented:
Mmmh. Wie funkioniert das denn als Plugin?
Da ich zwei WR habe muss ich die mittels `calc` addieren...

andig commented:
das ist kein plugin sondern ein meter. ein sma plugin hatten wir bisher ja auch nciht. aber gute punkt :O

premultiply commented:
Mit
```yaml
meters:
- name: shm
  type: sma
  uri: 192.168.178.90
- name: pv-haus
  type: sma
  uri: 192.168.178.91
- name: pv-garage
  type: sma
  uri: 192.168.178.92
- name: speicher
  type: sma
  uri: 192.168.178.93
```

Erhalte ich nur
```
pi@evcc:~/evcc $ sudo ./evcc meter --log trace
[main  ] INFO 2021/06/13 12:41:32 evcc 0.56 (f9d7a15)
[main  ] INFO 2021/06/13 12:41:32 using config file /home/pi/evcc/evcc.yaml
[mqtt  ] INFO 2021/06/13 12:41:32 connecting evcc-618135615 at localhost:1883
[mqtt  ] DEBUG 2021/06/13 12:41:32 localhost:1883 connected
[sma   ] TRACE 2021/06/13 12:41:32 new receive channel for 192.168.178.90
[sma   ] TRACE 2021/06/13 12:41:32 send package to 192.168.178.90: [proto.GroupPacketEntry, proto.SmaNet2PacketEntry]
[sma   ] TRACE 2021/06/13 12:41:32 received package [proto.GroupPacketEntry, proto.SmaNet2PacketEntry]
[sma   ] TRACE 2021/06/13 12:41:32 new energy meter at 192.168.178.90 - Serial=%!s(uint32=xxxxxxxxxx)
[sma   ] TRACE 2021/06/13 12:41:32 new receive channel for 192.168.178.91
[sma   ] TRACE 2021/06/13 12:41:32 send package to 192.168.178.91: [proto.GroupPacketEntry, proto.SmaNet2PacketEntry]
[sma   ] TRACE 2021/06/13 12:41:32 received package [proto.GroupPacketEntry, proto.SmaNet2PacketEntry]
[sma   ] TRACE 2021/06/13 12:41:32 new inverter at 192.168.178.91 - Serial=%!s(uint32=xxxxxxxxxx)
[sma   ] TRACE 2021/06/13 12:41:32 login for 192.168.178.91:9522
[sma   ] TRACE 2021/06/13 12:41:32 send package to 192.168.178.91: [proto.GroupPacketEntry, proto.SmaNet2PacketEntry]
[sma   ] TRACE 2021/06/13 12:41:32 received package [proto.GroupPacketEntry, proto.SmaNet2PacketEntry]
[main  ] FATAL 2021/06/13 12:41:32 cannot create meter 'pv-haus': cannot create meter 'sma': login failed
pi@evcc:~/evcc $
```

andig commented:
mach bitte ein neues ticket oder in https://github.com/andig/evcc/pull/1149 ergänzen
